### ISSUE TITLE
Latest update broke extension when no password was set

### ISSUE THREAD

**Issue by thallium205** [Authenticator-Extension/Authenticator#100]
Team, I woke up this morning to access my codes and was presented with this error:

![image](https://user-images.githubusercontent.com/872832/37857613-c17326e6-2eb9-11e8-8ee7-c7ab507ad35f.png)

I have never set a password and anything I type in results in "INVALID" or "ENCRYPTED" on the values now.  I cannot even export anything.  What's going on?

**Comment by qx2** [comment]
I have the exact same issue.

**Comment by Sneezry** [comment]
The message only show when you set a password and check remember passport option. I guess you may forget you had set a password someday. Because your browser remember the password, you were not asked to decrypt the secrets.

Is all your secret become to invalid or part of them?

**Comment by Brianetta** [comment]
I had a saved password, and lost some accounts as `Encrypted`.

**Comment by mymindstorm** [comment]
- Did any of you set passwords at any point in time?
- Have any of you used the extension for more than a month?
- Do you have any messages when you follow [these steps](https://github.com/Authenticator-Extension/Authenticator/wiki/Getting-Debugging-Information#chrome)

If you are sure you haven't set a password before, what is the result of pasting 
```javascript
console.log(typeof localStorage.encodedPhrase)
```
into the console?

**Comment by thallium205** [comment]
> Did any of you set passwords at any point in time?

No, never.

> Have any of you used the extension for more than a month?

I have used it for over 2 years.

> Do you have any messages when you follow these steps

Yes I receive:
```
Error in response to storage.get: TypeError: Cannot read property 'secret' of undefined
    at Object.chrome.storage.sync.get [as callback] (chrome-extension://bhghoamapcdpbohphigoooaddinpkbai/build/models/storage.js:170:75)
    at Promise (chrome-extension://bhghoamapcdpbohphigoooaddinpkbai/build/models/storage.js:123:41)
    at Function.<anonymous> (chrome-extension://bhghoamapcdpbohphigoooaddinpkbai/build/models/storage.js:121:20)
    at chrome-extension://bhghoamapcdpbohphigoooaddinpkbai/build/models/otp.js:12:71
    at __awaiter (chrome-extension://bhghoamapcdpbohphigoooaddinpkbai/build/models/otp.js:8:12)
```

> If you are sure you haven't set a password before, what is the result of pasting

```
console.log(typeof localStorage.encodedPhrase)
undefined
undefined
```

**Comment by mymindstorm** [comment]
@thallium205 run this, **do not paste anything from this here**
```
chrome.storage.sync.get(data=>console.log(JSON.stringify(data,null,2)))
```
there should be something like this repeated a few times:
```json
  "d41d8cd98f00b204e9800998ecf8427e": {
    "account": "",
    "counter": 0,
    "encrypted": false,
    "hash": "d41d8cd98f00b204e9800998ecf8427e",
    "index": 1,
    "issuer": "",
    "secret": "UIB4JSQOMILON6N6PXACWJGUAQUUFIBN",
    "type": "totp"
  }
```
In any of them after the part that says `"secret":` does it have text like above example or is it just `""`?

**Comment by thallium205** [comment]
@mymindstorm All objects in the map contain a value for key `secret` however they are encrypted and the flag `encrypted` is `true`.

**Comment by thallium205** [comment]
I should note that when I was first confronted with this error two days ago, I frantically tried to set a password so I could get passed it.  However, even after setting it, all values are stuck with "Encrypted".

**Comment by dankac** [comment]
So there is no chance to reset password  sneezry? I thought You can do it remotely no?

**Comment by Sneezry** [comment]
@dankac No. Your data is safe, and I have no access to your data.

**Comment by Sneezry** [comment]
@thallium205 I wonder if there is any space in your secret.

If this issue is caused by space in secret, your data is safe, and it should work in v5.0.9, which will release in this weekend.

**Comment by TheAhmadOsman** [comment]
Hey, I got a similar issue, except that even when I enter the password, not all my accounts get decrypted. Instead of getting the 2-steps code I am getting "Encrypted" for a few of my accounts.

**Comment by Sneezry** [comment]
@Ahmad-Magdy-Osman I wonder if there is any space in your secret.

If this issue is caused by space in secret, your data is safe, and it should work in v5.0.9, which will release in this weekend.

**Comment by TheAhmadOsman** [comment]
@Sneezry there is no space in neither my password nor my secret for those accounts that are saying "Encrypted". However, there is a "&nbsp;" in the names of the "issuer" field. That might be it?

**Comment by Sneezry** [comment]
@Ahmad-Magdy-Osman I mean the secret may contains spaces. If you copy it from a web page, it may contains spaces. However, it should work again in the next release.

**Comment by thallium205** [comment]
All my accounts, even after providing the password I created hastily, remain "Encrypted".

**Comment by Sneezry** [comment]
@thallium205 Then I guess you type incorrect password

**Comment by tanhaa** [comment]
I'm having a similar issue, I don't remember ever setting a password.  After trying several times in "settings", it offered me a chance to set a new password and by habit I put spaces in there.  SO now if I enter this password, everything still shows as "Encrypted".  Has a new release been pushed to handle that?  
```
Error in response to storage.get: TypeError: Cannot read property 'secret' of undefined
    at Object.chrome.storage.sync.get [as callback] (chrome-extension://bhghoamapcdpbohphigoooaddinpkbai/build/models/storage.js:170:75)
    at Promise (chrome-extension://bhghoamapcdpbohphigoooaddinpkbai/build/models/storage.js:123:41)
    at Function.<anonymous> (chrome-extension://bhghoamapcdpbohphigoooaddinpkbai/build/models/storage.js:121:20)
    at chrome-extension://bhghoamapcdpbohphigoooaddinpkbai/build/models/otp.js:12:71
    at __awaiter (chrome-extension://bhghoamapcdpbohphigoooaddinpkbai/build/models/otp.js:8:12)
```
````
console.log(typeof localStorage.encodedPhrase)

VM739:1 undefined
````

**Comment by Sneezry** [comment]
@tanhaa It seems your data is broken. Open popup page, right click and select Inspect, switch to Console tab, copy and paste code below, then hit Enter:

```
chrome.storage.sync.get(d=>{for(const k of Object.keys(d)){if(d[k].secret){d[k].secret='SECRET'}}console.log(JSON.stringify(d,null,2))})
```

What is the result you can see?

**Comment by Sneezry** [comment]
@tanhaa emmm, I find we have a bug related with your issue, and I have fixed it in the recent code. However, I still need you provide the result with running code above to help you to recover your data.

**Comment by tanhaa** [comment]
So i see all the entries with "SECRET" everywhere. 
(thanks for the quick reply)

**Comment by tanhaa** [comment]
I have pressed backup data in the past as well, where would that data be backed up?

**Comment by Sneezry** [comment]
@tanhaa I just want to confirm if your data is broken or it is caused by a bug. All secrets are replaced with "SECRET" to protect your privacy. You can post data you got from console here safely.

**Comment by tanhaa** [comment]
yes, I have an array of entries looking like this: 
```
 "178c4876cb376c687651229738c44463": {
    "account": "mygmailaccnt",
    "counter": 0,
    "encrypted": true,
    "hash": "178c4876cb376c687651229738c44463",
    "index": 5,
    "issuer": "Google",
    "secret": "SECRET",
    "type": "totp"
  },
```

**Comment by DtheRock** [comment]
Hi, 

I have exactly the same issue. 
I do not remember to have set a password in the past, however, when I add a password it decrypts some of the accounts and not all of them.

**Comment by Sneezry** [comment]
@DtheRock what's the version of the extension?

**Comment by tanhaa** [comment]
Is there anything that can be done to access my codes again?

**Comment by Sneezry** [comment]
@tanhaa first thing I need verify is if the extension update to 5.0.9, and if error remains in popup page. I notice you have set a password which contains only space. You should set password to empty to remove it. If your extension updates to 5.0.9, and all your entries display Encrypted, I believe you input an incorrect password. If you forget the correct password, I cannot help you to recover your data. If you have a backup, you can safely remove everything, and import the backup.

**Comment by DtheRock** [comment]
5.0.9

**Comment by Sneezry** [comment]
@DtheRock so there are still some accounts display Encrypted even if you input correct password, right?

**Comment by DtheRock** [comment]
Exactly

**Comment by Sneezry** [comment]
@DtheRock open popup page, input the correct password, right click, select Inspect, switch to Console tab, copy and paste code below:

```
chrome.storage.sync.get(d=>{for(let k of Object.keys(d)){let e=d[k];if(!e.encrypted)return;try{let s=CryptoJS.AES.decrypt(e.secret,document.cookie.split('=')[1]).toString(CryptoJS.enc.Utf8);if(!/^[a-z0-9=]+$/i.test(s)){let m=s.match(/[^a-z0-9=]/ig);console.log(k,m)}}catch(err){console.log(k,'error pass')}}})
```

Press enter, show me the result.

**Comment by DtheRock** [comment]
0f8d24dc5d36dd1afae011117e089dd0 null
VM101:1 106b5d298342f76587551814a9e0378d null
VM101:1 14e07a166a1eb4e21f03a047453fb358 null
VM101:1 35aea4d368e88d8a20dd5d48fb09b45d null
VM101:1 4713233e3667dc696e1cbb091b3f3d73 null
VM101:1 47246461f25293fa85e7ef781758e0b0 null
VM101:1 63fe2c8f12d445b3f8e717132dcd9ec1 error pass
VM101:1 64a10ea4d609c626197eeb31521f2aec error pass
VM101:1 82261ddeed6811d398f7c7c3cc6ebb8b error pass
VM101:1 a3c895c2681a78fb265a6af11f9ac9cd null
VM101:1 a8700539023da6cb02c3d1992562b903 null
VM101:1 bc1b4f33f7caf0271f31163c9b260efc null
VM101:1 def09fa4a833717d38298eb37c861af8 null
VM101:1 e4d539031af464b44c937870ed11fa21 null
VM101:1 e9259bbd6a8d0afc2c44f46548765679 null
VM101:1 f3013572622f3fa4c1aa0aa588476c47 null
VM101:1 undefined null

**Comment by Sneezry** [comment]
Three of your accounts cannot be decrypted with your current password. Could you provide more details about those three accounts? Such as are they added from another PC and synced to your current PC? When did you add them? What is the last time they worked?

**Comment by DtheRock** [comment]
These accounts were added a couple of months ago and they worked until this bug came up. 
They have not been imported from any other device or browser.

**Comment by Sneezry** [comment]
@DtheRock I have opened a new issue to track your issue, let's discuss there https://github.com/Authenticator-Extension/Authenticator/issues/113

**Comment by tanhaa** [comment]
My current version is 5.0.8.  

My password that was setup contains spaces, I set it up only after the app started prompting for password and suddenly allowed me to set a password.  I don't remember setting up a password when I had installed the app a long while ago.  

But now if I try to change the password, the extension keeps telling me that I must enter the password.  It doesn't accept the password it asked me to set after I couldn't access them anymore... not sure this is correct behaviour.   Where is the backup stored normally? 

This is what I see: 
<img width="324" alt="screen shot 2018-03-31 at 7 05 04 pm" src="https://user-images.githubusercontent.com/2850840/38168308-7a38a900-3516-11e8-97fd-a2916d9601a1.png">

It's weird cuz this is the password I set when the app wasn't working after the recent update.

**Comment by tanhaa** [comment]
I have updated the version to 5.0.9.  I am still having the same issue, app keeps telling my "ENCRYPTED".

**Comment by tanhaa** [comment]
@Sneezry:  by the way, I get this when I use the command you provided to the @DtheRock 
```
178c4876cb376c687651229738c44463 null
VM852:1 2698a64a6c408edcb58d672df408e676 null
VM852:1 41724294df8372350514ebe10a3cc25c error pass
VM852:1 8fa1a374d4e70052f3567286f20a4860 null
VM852:1 a6130ccb87f31ffa13ceb7591c42ec16 null
VM852:1 ce419e7be56ac0042e1fe00e31df7506 error pass
VM852:1 ddda29a9921706dc7925a6b455951732 null
VM852:1 def09fa4a833717d38298eb37c861af8 null
VM852:1 undefined null
```

**Comment by Sneezry** [comment]
@tanhaa Unfortunately, two of your accounts are locked with different passwords from the result I learnt. You need remove them manually and import them from backups again.

The extension won't backup data automatically, you need export backup yourself from setting menu (settings -> Export / Import). If you do not have backups, follow https://github.com/Authenticator-Extension/Authenticator/wiki/Lost-Codes to recover your account.

**Comment by tanhaa** [comment]
I can't seem to backup, export, import, etc from the menu, it keeps telling me I need to decrypt all my accounts, however, it does not decrypt  @anything! I wish I could just revert to a version that was working!  In addition, I remember pressing "backup" in the past, I just don't remember where it backs these codes up. 

By the way, not sure why two of them showed "error pass" and not sure how the extension can lock account with different passwords, there's obviously a bug in how this was handled.  Only one of the accounts in the popup shows "Invalid".  I'll remove that account and see if I can recover the other ones.  I have not been able to decrypt anything yet, even those that don't show "error pass"

EDIT:  I removed two accounts that showed up as invalid, the others are still not decrypting.  They continue showing up as "Encrypted" and any attempt to decrypt them doesn't work.  
```
chrome.storage.sync.get(d=>{for(let k of Object.keys(d)){let e=d[k];if(!e.encrypted)return;try{let s=CryptoJS.AES.decrypt(e.secret,document.cookie.split('=')[1]).toString(CryptoJS.enc.Utf8);if(!/^[a-z0-9=]+$/i.test(s)){let m=s.match(/[^a-z0-9=]/ig);console.log(k,m)}}catch(err){console.log(k,'error pass')}}})

undefined
VM1517:1 178c4876cb376c687651229738c44463 null
VM1517:1 2698a64a6c408edcb58d672df408e676 null
VM1517:1 41724294df8372350514ebe10a3cc25c null
VM1517:1 8fa1a374d4e70052f3567286f20a4860 null
VM1517:1 a6130ccb87f31ffa13ceb7591c42ec16 null
VM1517:1 ddda29a9921706dc7925a6b455951732 null


console.log(typeof localStorage.encodedPhrase)
VM1519:1 undefined
undefined
```

**Comment by tanhaa** [comment]
Here's how I got to the "Set password" portion, even if I had never set a password in the first place and the extension kept asking me for a password:
When you click on the extension Icon, it asks for a password: 
If you click OK, the next screen shows all your codes and "Invalid" where the code would usually be. 
Click on the settings Icon
Click on security
and It asks you to set a new password
<img width="361" alt="screen shot 2018-04-01 at 8 56 09 pm" src="https://user-images.githubusercontent.com/2850840/38179261-2e48161e-35ef-11e8-8ec1-2109657693d6.png">

so I guess that's how I got to setting up a password a few days ago. 
Now, when I tried to use that password, all i see is "Encrypted". 
After getting to that screen above, I left both password fields empty and apparently the "success" was with an empty password.  Did the password get changed? Apparently not, because trying to enter the app with the "empty password" i just setup, gives me "Invalid" again for codes everywhere. 

So I enter again with the password I had setup a few days ago, probably by doing this route as I had never setup a password on this app before (I installed this app somewhere in July/August last year I believe ) and the entries show up as "Encrypted" again, but I can't decrypt any of these.

**Comment by Sneezry** [comment]
@tanhaa When the extension asks password, you just leave the input box blank and click OK button, and all codes are shown as Invalid, right?

**Comment by tanhaa** [comment]
@Sneezry yes.

**Comment by Sneezry** [comment]
@tanhaa You must provide a password. Leave the input box blank will cause errors. I understand you are confused with the password, and you believe you had never set a password. But code won't lie.

In some previous versions, if you don't set a password, the extension will always show a warning to ask you to set a password in security menu, and I guess you set the password at that time, and you just forget you had set it, because the extension remember the password locally for you.

Now your account are locked with a password you set, you have to remember it. The password should be the one you saw a warning says your password is stored locally, and you should change it immediately.

Also, if you had exported backup from an old version, you should save it into a txt file or in an email. Old version extension shows some code in a text box to ask you copy and save it somewhere, somewhere only you know.

**Comment by tanhaa** [comment]
@Sneezry I disagree that your code is not lying about me setting a password.  In fact: https://github.com/Authenticator-Extension/Authenticator/issues/113  you said yourself that the code asks for password even if you didn't set one.  I don't use this often, in fact it's rare that with google I have to enter a MFA code, for other accounts that I had 2FA on, I normally used CLI to access so it was rare that I needed to use the MFA code to enter an account.  I have been prompted in the past to "Backup" the codes, but never anything about setting a password.  I have tried all my regular passwords that I would use and none worked leading me to believe that my assumption is correct and I had never set a password on this.  I'll go ahead and start resetting my accounts with 2FA and use my phone again for this, but I have to say that your extension definitely has some issues. You need to not allow empty password inputs to go through - and then allow them to go to settings and change password, which apparently doesn't really work!  This is horrible UX.   So you say I don't remember the password, I leave the input blank, I go in the settings  - security and I change my password!  Extension never says this is not right.  Very confusing experience with your latest versions and I have used this extension, like I said, since July of last year.  

Thanks for your help.

**Comment by mymindstorm** [comment]
@tanhaa this bug is a bit of a pain since neither of us can reporduce and any evidence from further up the thread completely contradicts (error in response to storage.get and undefined variable) what the code should be doing. Normally the password screen would work as expected, but it's not a good idea to try to set or attempt to remove a password that should not be on the account in the first place since that will corrupt your data. 

> I remember pressing "backup" in the past, I just don't remember where it backs these codes up.

Check your downloads folder for a file named `authenticator.json`, if you can find it  delete all codes, reimport them, and see if the issue still presists. Since you have reset your codes we can dump the secret data from the extension to help pinpoint the issue if you can find it.

**Comment by Elcheer** [comment]
Hi, similar problem, but inputting anything in the password field puts me in the authenticator list. Everything Encrypted.

**Comment by reni10** [comment]
Same problem for me!

Account shows Encrypted even after entering the password!

When I run the code you provided in the console I just get an undefined message back and nothing else!

Can you please fix this I am on version 5.0.9.

**Comment by thallium205** [comment]
When I do not enter a password at all when prompted I receive this as a response to the above code snippet:

```
VM391:1 12bbe8672f461af30774e5811d10edbd null
VM391:1 1dce7599ca6525f652314fc126edcd8b error pass
VM391:1 283f29ebf775886f30415b17e01ef4eb null
VM391:1 3b90a8730b925b8bccf2fffd662edbce null
VM391:1 63f04fae07ed649ebfe3e72985586478 null
VM391:1 7fd306d9e7ce0e2260fdc0b491db4e75 null
```

However, when I type in the password I *know* I set when prompted to a few weeks back by this plugin, I get this as a result:

```
VM472:1 12bbe8672f461af30774e5811d10edbd null
VM472:1 1dce7599ca6525f652314fc126edcd8b null
VM472:1 283f29ebf775886f30415b17e01ef4eb null
VM472:1 3b90a8730b925b8bccf2fffd662edbce error pass
VM472:1 63f04fae07ed649ebfe3e72985586478 null
VM472:1 7fd306d9e7ce0e2260fdc0b491db4e75 null
```
My experience is nearly identical to that of @tanhaa.  Also @Sneezry to respond to your earlier comment I absolutely did not forget my password as I used a password manager when setting this.  Is there any hope?

**Comment by Sneezry** [comment]
@Elcheer @reni10 @thallium205 If all your accounts cannot be decrypted with any password, you have to recover your account from backups. If a part of your accounts cannot be decrypted, you need remove them from the extension and recover them from backups.

As @mymindstorm said, this bug cannot reproduce on our side, it is very annoying...

**Comment by Elcheer** [comment]
Importing whatever authenticator.json I have closes the import window and does nothing.

**Comment by mymindstorm** [comment]
@Elcheer the window only closes if the import was successful. First of all remove all codes and then try and import again, if it is still is broken after that then is the word after `encrypted:`on a broken code `true`?

**Comment by mymindstorm** [comment]
To see this you may nedd to open `authenticator.json` in a text editor

**Comment by reni10** [comment]
There is something really wrong with this add-on now!

I have setup another brand new account and it worked for a few days but now has gone back to saying "Encrypted" and even though I enter in the exact password it still never changes from "Encrypted" so can you please look into this asap?

**Comment by Geograph-us** [comment]
+1 same issue

**Comment by mymindstorm** [comment]
It's been three months and no new info on this has surfaced. If anyone can reproduce this using the latest stable version of Chrome and the extension then please open a new report.
